File: demonstrations/universal_testing_automation_20251012_171832/calculator.ts
export class Calculator {
    add(a: number, b: number): number {
        return a + b;
    }

    subtract(a: number, b: number): number {
        return a - b;
    }

    multiply(a: number, b: number): number {
        return a * b;
    }

    divide(a: number, b: number): number {
        if (b === 0) {
            throw new Error("Division by zero");
        }
        return a / b;
    }

    power(base: number, exponent: number): number {
        return Math.pow(base, exponent);
    }

    factorial(n: number): number {
        if (n < 0) {
            throw new Error("Negative factorial");
        }
        return n === 0 ? 1 : n * this.factorial(n - 1);
    }

    fibonacci(n: number): number {
        if (n < 0) {
            throw new Error("Negative fibonacci");
        }
        if (n <= 1) {
            return n;
        }
        return this.fibonacci(n - 1) + this.fibonacci(n - 2);
    }
}

export class AdvancedCalculator extends Calculator {
    squareRoot(x: number): number {
        if (x < 0) {
            throw new Error("Negative square root");
        }
        return Math.sqrt(x);
    }

    logarithm(x: number, base: number = Math.E): number {
        if (x <= 0) {
            throw new Error("Non-positive logarithm");
        }
        return Math.log(x) / Math.log(base);
    }
}

File: demonstrations/universal_testing_automation_20251012_171832/calculator.py
class Calculator:
    def add(self, a, b):
        return a + b

    def subtract(self, a, b):
        return a - b

    def multiply(self, a, b):
        return a * b

    def divide(self, a, b):
        if b == 0:
            raise ZeroDivisionError("Division by zero")
        return a / b

    def power(self, base, exponent):
        return base ** exponent

    def factorial(self, n):
        if n < 0:
            raise ValueError("Negative factorial")
        return 1 if n == 0 else n * self.factorial(n - 1)

    def fibonacci(self, n):
        if n < 0:
            raise ValueError("Negative fibonacci")
        if n <= 1:
            return n
        return self.fibonacci(n - 1) + self.fibonacci(n - 2)
